### PR TITLE
Added persister for node_agent & some basic node inventory.

### PR DESCRIPTION
**Pull request by nnDarshan**
This patch adds persister for node_agent and also adds few basic
inventory data collection like os details, cpu details, memory
details.

NOTE: This is still work under progress patch and not verified

Signed-off-by: nnDarshan darshan.n.2024@gmail.com

### CODE REVIEW

**Comment by nnDarshan** [comment]
@r0h4n have pushed the WIP patch for adding persistence for node_agent. Please have a look. This is not tested.

**Comment by nnDarshan** [comment]
@r0h4n @shtripat @nthomas-redhat Please review

File: tendrl/node_agent/manager/manager.py [hunk 1]
```diff
@@ -3,13 +3,57 @@
...
+        self._complete.set()
+
+    def _run(self):
+        LOG.info("%s running" % self.__class__.__name__)
+
+        while not self._complete.is_set():
+            try:
+                gevent.sleep(3)
+                pull_hardware_inventory.write_node_inventory(
+                    "/tmp/tendrl-node-inventory.json"
```
Comment (shtripat) [inline tendrl/node_agent/manager/manager.py]: Is it good to use /tmp for these kind of work? May be some tendrl specific dir?

**Comment by r0h4n** [comment]
Few changes required
1) On installation of node_agent on unmanaged node, create uuid.uuid4() and save at /etc/tendrl/node_agent_key
1) Register the node with the {node_agent_key : fqdn_hostname} in a dict at /nodes/metadata  (eg {nodes: [{node_agent_key: fqdn_hostname_of_node}], metadata1: value1}

Also, please provide a demo data dump or etcd url with these changes

File: etc/tendrl_definitions_node_agent.yaml [hunk 2]
```diff
@@ -20,26 +20,23 @@ object_details:
     enabled: True
```
Comment (r0h4n) [inline etc/tendrl_definitions_node_agent.yaml]: Need "package" object (to fetch list of rpms installed, yum repos etc) and "service" object (to fetch list of running systemd services etc)

File: tendrl/node_agent/persistence/servers.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+from tendrl.bridge_common.etcdobj import fields
+
+
+class Cpu(EtcdObj):
+    """A table of the cpu, lazily updated
+
+    """
+    __name__ = 'nodes/%s/cpu'
+
+    node_uuid = fields.StrField("node_uuid")
```
Comment (nthomas-redhat) [inline tendrl/node_agent/persistence/servers.py]: As we discussed, the field type definitions should come from yaml file. I understand that, it depends on the another patch Tim is working on. This is fine for now, but needs to be implemented that way

File: tendrl/node_agent/manager/manager.py [hunk 4]
```diff
@@ -19,22 +82,114 @@ def __init__(self):
...
 
 
 def main():
     log.setup_logging()
+    # Configure a uuid on the node, so that this can be used by Tendrl for
+    # uniquely identifying the node
+    try:
+        configure_tendrl_uuid()
+        LOG.info("Verified that node uuid exists at"
+                 " /etc/tendrl/node_agent_key")
```
Comment (r0h4n) [inline tendrl/node_agent/manager/manager.py]: "node_agent_key_$timestamp"

File: tendrl/node_agent/manager/manager.py [hunk 5]
```diff
@@ -19,22 +82,114 @@ def __init__(self):
...
+                    node_uuid=raw_data["node_uuid"],
+                )
+            )
+
+
+def configure_tendrl_uuid():
+    # check if valid uuid is already present in tendrl_node_id_file.
+    # if not present generate one and update the file
+    try:
+        with open(TENDRL_NODE_ID_FILE) as f:
```
Comment (r0h4n) [inline tendrl/node_agent/manager/manager.py]: timestamp required while updating, also delete the older node_agent_key_$oldtimestamp

at any given point in time, there should be only one /etc/tendrl/node_agent_key_$latesttimestamp

File: tendrl/node_agent/manager/pull_hardware_inventory.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
+from command import Command
+import platform
+TENDRL_NODE_ID_FILE = "/etc/tendrl/node_agent_key"
```
Comment (nnDarshan) [inline tendrl/node_agent/manager/pull_hardware_inventory.py]: Done

File: tendrl/node_agent/manager/manager.py (hunk 4 above)
Comment (nnDarshan) [inline tendrl/node_agent/manager/manager.py]: Done

File: tendrl/node_agent/manager/manager.py (hunk 5 above)
Comment (nnDarshan) [inline tendrl/node_agent/manager/manager.py]: Done

File: tendrl/node_agent/manager/manager.py [hunk 7]
```diff
@@ -3,11 +3,74 @@
...
+from tendrl.node_agent.persistence.servers import Cpu
+from tendrl.node_agent.persistence.servers import Memory
+from tendrl.node_agent.persistence.servers import NodeMetadata
+from tendrl.node_agent.persistence.servers import Os
+import time
+import uuid
+
 LOG = logging.getLogger(__name__)
+HARDWARE_INVENTORY_FILE = "/etc/tendrl/tendrl-node-inventory.json"
+TENDRL_NODE_ID_FILE = "/etc/tendrl/node_agent_key"
```
Comment (nnDarshan) [inline tendrl/node_agent/manager/manager.py]: Done

File: tendrl/node_agent/manager/manager.py [hunk 8]
```diff
@@ -3,11 +3,74 @@
...
 import gevent.event
 import gevent.greenlet
-
+import json
+import pull_hardware_inventory
 from rpc import EtcdThread
 from tendrl.node_agent import log
 
+from tendrl.node_agent.persistence.persister import Persister
+from tendrl.node_agent.persistence.servers import Cpu
```
Comment (nnDarshan) [inline tendrl/node_agent/manager/manager.py]: Done

File: tendrl/node_agent/manager/manager.py [hunk 9]
```diff
@@ -158,34 +168,46 @@ def on_pull(self, raw_data):
...
 def configure_tendrl_uuid():
-    # check if valid uuid is already present in tendrl_node_id_file.
+    # check if valid uuid is already present in node_agent_key
     # if not present generate one and update the file
+    file_list = []
+    for f in os.listdir(TENDRL_CONF_PATH):
+        if f.startswith("node_agent_key_"):
+            file_list.append(f)
+    if len(file_list) == 0:
+        with open(NODE_AGENT_KEY, 'w') as f:
```
Comment (nthomas-redhat) [inline tendrl/node_agent/manager/manager.py]: handle the exceptions here

Comment (nnDarshan) [inline tendrl/node_agent/manager/manager.py]: this is handled in the caller

**Pull request merged**